### PR TITLE
Fixing behaviour of T Model with zero DBH values

### LINKED ISSUE #524: Behaviour of T Model allometry and allocation with zero DBH values
Currently, if the DBH ($D$) of a stem is zero, then in the `StemAllometry` class:

* the crown fraction ($f_c$) is NaN, so...
* the sapwood mass (which uses $f_c$) is NaN 

And in the `StemAllocation` class:

* Consequently, sapwood respiration is NaN and as a result NPP and all the delta variables (diameter, mass, foliage) are NaN.

### PR DESCRIPTION
**Pull request by davidorme**
# Description

> [!Tip]
> OK - this is ready for review now.

Currently, if the DBH ($D$) of a stem is zero, then in the `StemAllometry` class:

* the crown fraction ($f_c$) is NaN, so...
* the sapwood mass (which uses $f_c$) is NaN 

And in the `StemAllocation` class:

* Consequently, sapwood respiration is NaN and as a result NPP and all the delta variables (diameter, mass, foliage) are NaN.

That leads to all sorts of unexpected behaviour. We _could_ make more sensible predictions when $D=0$ - which is what the first iteration of this PR did - but it basically doesn't make sense to make predictions when $D=0$.  That's actually a bit more work because we currently allow zero so this PR:

* Adds a new `_enforce_positive_sizes` helper function to `pyrealm.demography.tmodel` 
* Calls that function from within _nearly_ all of the validation sections of the T model functions. We don't call it on the reproductive mass functions because that is an experimental extension that we want to be able to set to zero.
* Add tests that zero and negative values are caught correctly.
* Updates tests to pass

The PR also: 

* Makes it a ValueError to create `StemAllometry` or `Cohorts` with negative DBH - this guarantees positive sizes and allows the per function validation to be skipped.
* Makes `Cohorts.dbh_values` a property raises ValueError when set with negative DBH values.

Fixes #524 

## Type of change

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pyrealm/demography/community.py [hunk 1]
```diff
@@ -173,15 +173,16 @@ class Cohorts(PandasExporter, CohortMethods):
     count_attr: ClassVar[str] = "n_cohorts"
 
     # Instance attributes
-    dbh_values: NDArray[np.float64]
     n_individuals: NDArray[np.int_]
     pft_names: NDArray[np.str_]
     _cohort_id: NDArray[np.str_] = field(init=False)
+    _dbh_values: NDArray[np.float64] = field(init=False)
```
Comment (MarionBWeinzierl) [inline pyrealm/demography/community.py:179]: Probably just me as I am not working in the field, but for me and other field-outsiders of the code it would help to add a short comment what `pft` or and `dbh` is (ideally, the variable names would be clear, but I assume these are abbreviations which are commonly used in the field?).

File: pyrealm/demography/community.py [hunk 2]
```diff
@@ -173,15 +173,16 @@ class Cohorts(PandasExporter, CohortMethods):
...
     n_cohorts: int = field(init=False)
+    dbh_values: InitVar[NDArray[np.float64]]
```
Comment (MarionBWeinzierl) [inline pyrealm/demography/community.py:181]: I am a bit confused here why we need `dbh_values` as well as `_dbh_values`.

**Review by MarionBWeinzierl — COMMENTED** [review]
I am now really curious what DBH means... :D

File: pyrealm/demography/community.py (hunk 1 above)
Comment (davidorme) [reply pyrealm/demography/community.py:179]: They are "plant functional type" and "diameter at breast height" - which are just a bit long winded to spell out so those abbreviations are commonly used.

File: pyrealm/demography/community.py (hunk 2 above)
Comment (davidorme) [reply pyrealm/demography/community.py:181]: It's a property thing: the user provides `dbh_values` to create the instance, which is then passed through the validation in the property setter and stored in `_dbh_values`. When a user accesses `dbh_values` the property decorator function returns the contents of the validated  `_dbh_values`.

It's a relatively common Python idiom for controlling class attributes (e.g. https://realpython.com/python-property/), but just looks a bit weird within a dataclass because some of the class attribute creation is concealed within the `@dataclass`. We could just make it a normal class for clarity, but this works as intended.

File: pyrealm/demography/community.py (hunk 1 above)
Comment (davidorme) [reply pyrealm/demography/community.py:179]: It's not a direct solve for reading the code, but I've added #526 to add a glossary.

File: pyrealm/demography/tmodel.py [hunk 3]
```diff
@@ -500,9 +549,13 @@ def calculate_reproductive_tissue_respiration(
         validate: Boolean flag to suppress argument validation
     """
     if validate:
+        size_args = {"reproductive_tissue_mass": reproductive_tissue_mass}
         _validate_demography_array_arguments(
-            trait_args={"resp_rt": resp_rt},
-            size_args={"reproductive_tissue_mass": reproductive_tissue_mass},
+            trait_args={"resp_rt": resp_rt}, size_args=size_args
+        )
+        _enforce_positive_sizes(
```
Comment (sjavis) [inline pyrealm/demography/tmodel.py]: A later function (calculate_reproductive_tissue_turnover) allows a reproductive tissue mass of zero. Should that be allowed here?

File: pyrealm/demography/tmodel.py [hunk 4]
```diff
@@ -785,6 +855,13 @@ def calculate_growth_increments(
     * the crown area ratio (:math:`c`), and
     * the ratio of fine root mass to leaf area (:math:`\zeta`).
 
+    The value of :math`\Delta D` is unstable when :math:`D = 0` and hence :math:`H = 0`
+    and the rates of change in stem and foliar mass are also zero. If :math:`P_{net} - T
+    = 0` then :math`\Delta D` is undefined, otherwise :math`\Delta D = \pm \inf`
+    depending on whether then turnover costs exceed the available NPP. Under these
+    conditions, this function explicitly sets :math`\Delta D = 0`: **stems with zero
```
Comment (sjavis) [inline pyrealm/demography/tmodel.py]:
```suggestion
    The value of :math:`\Delta D` is unstable when :math:`D = 0` and hence :math:`H = 0`
    and the rates of change in stem and foliar mass are also zero. If :math:`P_{net} - T
    = 0` then :math:`\Delta D` is undefined, otherwise :math:`\Delta D = \pm \inf`
    depending on whether then turnover costs exceed the available NPP. Under these
    conditions, this function explicitly sets :math:`\Delta D = 0`: **stems with zero
```

**Review by sjavis — APPROVED** [review]
Overall this looks good to me aside from a couple of comments.
